File: intermediate/actions_py/actions_py/count_until_server.py
#! /usr/bin/env python3

import time
import rclpy
from rclpy.executors import MultiThreadedExecutor
from rclpy.node import Node
from rclpy.action import ActionServer
from rclpy.action.server import (
    ServerGoalHandle, 
    GoalResponse, 
    CancelResponse)

from rclpy.callback_groups import ReentrantCallbackGroup
from ros2_custom_interfaces.action import CountUntil

class CountUntilServerNode(Node):
    def __init__(self):
        """
        Initializes the CountUntilServerNode class.

        This constructor sets up the action server for the CountUntil action.
        It creates an instance of the ActionServer class, specifying the action type,
        server name, and callback functions for goal handling, execution, and cancellation.
        It also logs a message indicating that the action server has started and is waiting for goals.

        Parameters:
        None

        Returns:
        None
        """
        super().__init__('count_until_server_node')

        self.count_until_server = ActionServer(
            self,
            CountUntil,
            "count_until",
            execute_callback=self.execute_callback,
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback,
            callback_group=ReentrantCallbackGroup(),
        )

        self.get_logger().info("Action server started. Waiting for goal.")

    def execute_callback(self, goal_handle: ServerGoalHandle):
        """
        Executes the action server's main logic.

        This function is called when a new goal is received by the action server.
        It counts up to the target number at the specified period, publishing feedback
        and checking for cancellation requests.

        Parameters:
        goal_handle (ServerGoalHandle): The handle for the received goal. Contains the goal request.

        Returns:
        CountUntil.Result: The result of the action, containing the reached number.
        """
        # Get request from goal
        target_number = goal_handle.request.target_number
        period = goal_handle.request.period

        # Feedback 
        feedback = CountUntil.Feedback()
        result = CountUntil.Result()

        # Action logic
        # Wait for user input to start counting up to the goal number
        self.get_logger().info(f"Received goal: {target_number}")
        counter = 0
        # Execute the action
        for i in range(target_number):
            if goal_handle.is_cancel_requested:
                self.get_logger().info("Goal Cancelled!")
                goal_handle.canceled()
                result.reached_number = counter
                return result

            counter = counter + 1
            feedback.current_number = counter
            goal_handle.publish_feedback(feedback)
            self.get_logger().info(f"Counting up to {target_number} - {counter}")
            time.sleep(period)

        # Once done, Set goal final state
        goal_handle.succeed()
        # Send result
        self.get_logger().info(f"Goal reached: {target_number}")
        result = CountUntil.Result()
        result.reached_number = counter

        return result
    
    def goal_callback(self, goal_request: CountUntil.Goal):
        """
        Callback function for handling incoming goal requests.

        This function is called whenever a new goal is received by the action server.
        It validates the goal request and either accepts or rejects it based on the target number.

        Parameters:
        goal_request (CountUntil.Goal): The incoming goal request containing the target number and period.

        Returns:
        GoalResponse: The response indicating whether the goal was accepted or rejected.
        """
        self.get_logger().info(f"Received goal: {goal_request.target_number}")

        # Validate goal request
        if goal_request.target_number <= 0:
            self.get_logger().warn("Goal Rejected! Target number must be greater than 0") 
            return GoalResponse.REJECT

        self.get_logger().info("Goal Accepted!")
        return GoalResponse.ACCEPT
    
    def cancel_callback(self, goal_handle: ServerGoalHandle) -> CancelResponse:
        """
        Callback function for handling cancellation requests.

        This function is called whenever a goal cancellation request is received by the action server.
        It simply accepts the cancellation request and returns the appropriate response.

        Parameters:
        goal_handle (ServerGoalHandle): The handle for the goal that is being cancelled.

        Returns:
        CancelResponse: The response indicating whether the cancellation request was accepted.
        In this case, it always returns CancelResponse.ACCEPT.
        """
        return CancelResponse.ACCEPT


def main(args=None):
    """
    The main function initializes the ROS2 environment, creates an instance of the CountUntilServerNode,
    spins the node using a MultiThreadedExecutor, destroys the node, and shuts down the ROS2 environment.

    Parameters:
    args (list, optional): A list of command-line arguments to be passed to rclpy.init().
                           Defaults to None, which means no arguments are passed.

    Returns:
    None
    """
    rclpy.init(args=args)
    server_node = CountUntilServerNode()
    rclpy.spin(server_node, MultiThreadedExecutor())
    server_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: intermediate/actions_cpp_package/src/count_until_server.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include "ros2_custom_interfaces/action/count_until.hpp"

class CountUntilServer : public rclcpp::Node {

    public:
      /**
       * @brief Constructor for the CountUntilServer class. Initializes the ROS node and action server.
       *
       * This constructor creates an action server for the "count_until" action. It binds the goal,
       * cancel, and accepted callback functions to handle incoming goals and cancel requests.
       *
       * @param[in] Node::Node("count_until_server_node") The name of the ROS node.
       */
      CountUntilServer(): Node("count_until_server_node"){
        // Create an action server.
        count_until_server_ = rclcpp_action::create_server<ros2_custom_interfaces::action::CountUntil>(
            this,
            "count_until",
            std::bind(&CountUntilServer::goalCallback, this, std::placeholders::_1, std::placeholders::_2),
            std::bind(&CountUntilServer::cancelCallback, this, std::placeholders::_1),
            std::bind(&CountUntilServer::handleAcceptedCallback, this, std::placeholders::_1)
            );

        RCLCPP_INFO(this->get_logger(), "CountUntil Server is ready to accept goals.");
      }



    private:
      /**
       * @brief Callback function to handle incoming goal requests.
       *
       * This function is called when a new goal is received by the action server.
       * It determines whether the goal should be accepted or rejected.
       *
       * @param[in] uuid The unique identifier for the goal request.
       * @param[in] goal A shared pointer to the goal message containing the target number and period.
       * 
       * @return rclcpp_action::GoalResponse The response indicating whether the goal is accepted or rejected.
       *         In this implementation, the goal is always accepted and executed.
       */
      rclcpp_action::GoalResponse goalCallback(
          const rclcpp_action::GoalUUID &uuid, 
          std::shared_ptr<const ros2_custom_interfaces::action::CountUntil::Goal> goal){
        (void)uuid;
        (void)goal;
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
      }


      /**
       * @brief Callback function to handle cancel requests for goals.
       *
       * This function is called when a cancel request is received for a goal.
       * It determines whether the cancel request should be accepted or rejected.
       *
       * @param[in] goal_handle A shared pointer to the goal handle for which the cancel request is made.
       * 
       * @return rclcpp_action::CancelResponse The response indicating whether the cancel request is accepted or rejected.
       *         In this implementation, the cancel request is always accepted.
       */
      rclcpp_action::CancelResponse cancelCallback(
        const std::shared_ptr<rclcpp_action::ServerGoalHandle<ros2_custom_interfaces::action::CountUntil>> goal_handle
      ){
        (void)goal_handle;
        return rclcpp_action::CancelResponse::ACCEPT;
      }

      void handleAcceptedCallback(
        const std::shared_ptr<rclcpp_action::ServerGoalHandle<ros2_custom_interfaces::action::CountUntil>> goal_handle){
        executeGoal(goal_handle);
      }

      void executeGoal(
        const std::shared_ptr<rclcpp_action::ServerGoalHandle<ros2_custom_interfaces::action::CountUntil>> goal_handle
      ){
        int target_number = goal_handle->get_goal()->target_number;

        double period = goal_handle->get_goal()->period;

        RCLCPP_INFO(this->get_logger(), "Executing CountUntil action, target number: %d, period: %.2f seconds", target_number, period);
        int counter = 0;
        rclcpp::Rate loop_rate(1.0 / period);
        for (int i = 0;  i < target_number; i++)
        {
          counter++;
          RCLCPP_INFO(this->get_logger(), "Current count: %d", counter);
          loop_rate.sleep();
        }

        // Set final state and return
        auto result = std::make_shared<ros2_custom_interfaces::action::CountUntil::Result>();
        result->reached_number = counter;
        goal_handle->succeed(result);
      }

      rclcpp_action::Server<ros2_custom_interfaces::action::CountUntil>::SharedPtr count_until_server_;
};

int main(int argc, char ** argv){
  rclcpp::init(argc, argv);
  auto node = std::make_shared<CountUntilServer>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
